File: audit-platform/parser/tb_parser/normalize.py
"""Text hygiene for client-supplied account codes and names.

Two outputs, kept deliberately separate:

* ``clean_display`` — what auditors see and sign. The characters are the
  client's, minus the invisible ones that let a string *render* differently
  from its bytes (bidi embeddings/overrides/isolates, directional marks,
  zero-width characters, control characters). Nothing else is changed.

* ``normalize_for_matching`` — used only to compare names (mapping cascade).
  NFKC (folds Arabic presentation forms and full-width Latin), case-folding,
  ASCII digits, Arabic orthographic folding (alef/yaa/taa-marbuta/hamza
  carriers, tashkeel, tatweel), punctuation to spaces.

The TypeScript port (backend/src/modules/tb-ingestion/normalize.ts) must agree
byte-for-byte; both are tested against tests/golden/normalization.json.
"""

from __future__ import annotations

import re
import unicodedata

# Characters that change how text is displayed without being visible.
_INVISIBLE = {
    0x061C,  # ARABIC LETTER MARK
    0x200B,  # ZERO WIDTH SPACE
    0x200E, 0x200F,  # LRM, RLM
    0x202A, 0x202B, 0x202C, 0x202D, 0x202E,  # LRE, RLE, PDF, LRO, RLO
    0x2060,  # WORD JOINER
    0x2066, 0x2067, 0x2068, 0x2069,  # LRI, RLI, FSI, PDI
    0xFEFF,  # BOM / ZWNBSP
    0x00AD,  # SOFT HYPHEN
}
_ZW_JOINERS = {0x200C, 0x200D}  # meaningful in some scripts: kept for display only

_WS = re.compile(r"\s+", re.UNICODE)

_ARABIC_FOLD = {
    "أ": "ا", "إ": "ا", "آ": "ا", "ٱ": "ا",
    "ٲ": "ا", "ٳ": "ا",           # alef variants -> alef
    "ى": "ي", "ی": "ي", "ئ": "ي",  # alef maqsura, farsi yeh, yeh+hamza -> yeh
    "ؤ": "و",                               # waw+hamza -> waw
    "ة": "ه",                               # taa marbuta -> heh
    "ک": "ك",                               # keheh -> kaf
}


def _is_tashkeel(cp: int) -> bool:
    return 0x064B <= cp <= 0x065F or cp == 0x0670 or 0x06D6 <= cp <= 0x06ED


def clean_display(text: str) -> tuple[str, bool]:
    """Return (cleaned text, whether invisible characters were removed)."""
    out: list[str] = []
    removed = False
    for ch in text:
        cp = ord(ch)
        if cp in _INVISIBLE:
            removed = True
            continue
        cat = unicodedata.category(ch)
        if cat == "Cc":  # C0/C1 controls: tab/newline become spaces, the rest go
            if ch in "\t\n\r\v\f":
                out.append(" ")
            else:
                removed = True
            continue
        if cat in ("Zl", "Zp"):
            out.append(" ")
            continue
        out.append(ch)
    return _WS.sub(" ", "".join(out)).strip(), removed


def ascii_digits(text: str) -> str:
    """Map every Unicode decimal digit (Arabic-Indic, Eastern Arabic-Indic, ...) to ASCII."""
    return "".join(str(unicodedata.decimal(ch)) if unicodedata.category(ch) == "Nd" else ch for ch in text)


def normalize_for_matching(text: str) -> str:
    cleaned, _ = clean_display(text)
    s = unicodedata.normalize("NFKC", cleaned).casefold()
    s = ascii_digits(s)
    out: list[str] = []
    for ch in s:
        cp = ord(ch)
        if _is_tashkeel(cp) or cp == 0x0640 or cp in _ZW_JOINERS:  # harakat, tatweel, ZWNJ/ZWJ
            continue
        ch = _ARABIC_FOLD.get(ch, ch)
        cat = unicodedata.category(ch)
        out.append(ch if cat[0] in ("L", "N") else " ")
    return _WS.sub(" ", "".join(out)).strip()


# Cells beginning with these execute as formulas when a CSV/XLSX export is
# opened in a spreadsheet (OWASP CSV injection).
FORMULA_TRIGGERS = ("=", "+", "-", "@", "\t", "\r")


def looks_like_formula(text: str) -> bool:
    return text.startswith(FORMULA_TRIGGERS) and not _is_plain_negative_number(text)


_NEG_NUMBER = re.compile(r"^[-+]\s*[\d.,٠-٩۰-۹٫٬]+$")


def _is_plain_negative_number(text: str) -> bool:
    return bool(_NEG_NUMBER.match(text))

File: audit-platform/backend/src/modules/tb-ingestion/normalize.ts
/**
 * TypeScript port of parser/tb_parser/normalize.py. The parser normalises TB
 * account names in the sandbox; the backend must produce the SAME matching
 * form for the other side of each comparison (chart-of-accounts names, firm
 * rules). Both implementations are tested against
 * parser/tests/golden/normalization.json.
 */

const INVISIBLE = new Set([
  0x061c, 0x200b, 0x200e, 0x200f, 0x202a, 0x202b, 0x202c, 0x202d, 0x202e, 0x2060, 0x2066, 0x2067, 0x2068,
  0x2069, 0xfeff, 0x00ad,
]);
const ZW_JOINERS = new Set([0x200c, 0x200d]);

const ARABIC_FOLD: Record<string, string> = {
  'أ': 'ا', 'إ': 'ا', 'آ': 'ا', 'ٱ': 'ا', 'ٲ': 'ا', 'ٳ': 'ا',
  'ى': 'ي', 'ی': 'ي', 'ئ': 'ي',
  'ؤ': 'و',
  'ة': 'ه',
  'ک': 'ك',
};

const CONTROL = /\p{Cc}/u;
const LETTER_OR_NUMBER = /[\p{L}\p{N}]/u;
const DECIMAL_DIGIT = /\p{Nd}/u;
const WS = /\s+/gu;

function isTashkeel(cp: number): boolean {
  return (cp >= 0x064b && cp <= 0x065f) || cp === 0x0670 || (cp >= 0x06d6 && cp <= 0x06ed);
}

/** Remove characters that make text render differently from its bytes. */
export function cleanDisplay(text: string): { text: string; removed: boolean } {
  let out = '';
  let removed = false;
  for (const ch of text) {
    const cp = ch.codePointAt(0)!;
    if (INVISIBLE.has(cp)) { removed = true; continue; }
    if (CONTROL.test(ch)) {
      if ('\t\n\r\v\f'.includes(ch)) out += ' ';
      else removed = true;
      continue;
    }
    out += cp === 0x2028 || cp === 0x2029 ? ' ' : ch;
  }
  return { text: out.replace(WS, ' ').trim(), removed };
}

/**
 * Value of a Unicode decimal digit. Nd characters come in runs that start at a
 * zero and are whole multiples of ten long (e.g. the five consecutive sets of
 * mathematical digits), so the value is the offset from the run start mod 10.
 */
function digitValue(ch: string): string {
  const cp = ch.codePointAt(0)!;
  let start = cp;
  while (cp - start < 100 && DECIMAL_DIGIT.test(String.fromCodePoint(start - 1))) start -= 1;
  return String((cp - start) % 10);
}

export function asciiDigits(text: string): string {
  let out = '';
  for (const ch of text) out += DECIMAL_DIGIT.test(ch) ? digitValue(ch) : ch;
  return out;
}

/** Python str.casefold() for the scripts we handle: upper-then-lower expands ß -> ss, folds final sigma. */
function casefold(s: string): string {
  return s.toUpperCase().toLowerCase();
}

export function normalizeForMatching(text: string): string {
  const cleaned = cleanDisplay(text).text;
  const s = asciiDigits(casefold(cleaned.normalize('NFKC')));
  let out = '';
  for (let ch of s) {
    const cp = ch.codePointAt(0)!;
    if (isTashkeel(cp) || cp === 0x0640 || ZW_JOINERS.has(cp)) continue;
    ch = ARABIC_FOLD[ch] ?? ch;
    out += LETTER_OR_NUMBER.test(ch) ? ch : ' ';
  }
  return out.replace(WS, ' ').trim();
}

/** Account codes: invisible characters out, ASCII digits, trimmed, upper-case. */
export function normalizeCode(code: string): string {
  return asciiDigits(cleanDisplay(code).text).toUpperCase();
}
